OIDC SSO Parsing Tokens (#5447)
mendix/docs: content/en/docs/appstore/modules/oidc.md

### 1.1 Typical Usage Scenarios

* **B2C apps:** Your app is aimed at consumers who have an identity at a 'social IdP' which uses OIDC, such as Google. In this case your app will only delegate the authentication to the IdP, no further user information is available to the app.
* **B2E app:** Your app is aimed at your company's employees and you want these employees to sign in to your app using corporate credentials hosted by your identity provider (IdP) that supports the OIDC protocol. In this case your app may have its own logic to assign user roles or you may use authorization information from your IdP as provided to your app using an access token.
* **API consumption.** If your app makes calls to APIs of other services on behalf of your end-user, you can use the access token obtained via the “OIDC SSO” module. This scenario is not supported when using SAML SSO. This makes the OIDC SSO module suitable for Mendix customers using Mendix Data Hub.

> * **Authorizing access to a Mendix backend app.**  If you want to secure APIs for Mendix back-end apps using an access token, the app consuming your API can obtain that access token and pass it to your API in the authorization header. If your IDP produces access tokens that are JWTs, this brings the possibility of propagating the user and/or the user’s authorizations to your app and have user roles assigned based on the claims in the access token JWT. 

Reviewer: @MarkvanMents , I think in my draft it was more clear that this usage scenario is about the API (server side) and not so much about the client side of the process. can you improve wording to make that clear?
I thjink it may confuse the reader

Author: ```suggestion
* **Authorizing access to a Mendix back-end app.**  If you want to secure APIs in Mendix back-end apps using an access token, your API can use an access token passed by the calling app in the authorization header. If the access token is a JWT, your app can use the user and/or the user’s authorizations to assign user roles based on the claims in the access token JWT. 
```

Reviewer: 👍looks good to me :-)

* Uses the Authorization Code Grant flow to sign the end-user in via the browser.
* Uses the 'nonce' parameter to defend against replay attacks.
* Validates ID-token signatures.

> * It validates access tokens in one of two ways:

Reviewer 2: @MarkvanMents 
Can this be changed to "Validates access tokens while API authentication in one of two ways:" 
 since it is specific to API Authentication scenario only

>     * If the IdP supports token introspection, the OIDC module will introspect the access token to see if it is valid.

Reviewer: /introspect is an endpoint at the customer's IdP - I would like to have that more explicit.

Reviewer: This is part of the 'breaking change' as well:
- the microflows are organized differently so the custom one is simpler
- AT validation was something the microflow previously didn't do (correct Sharishma?)
- the validation of the AT via introspect (if possible) is added.

Reviewer 2: @JaapF 
We are using introspection endpoint for only for APIAuthentication. It will not play any role in Login flow via IDP.

To be more specific,there are two flows: 
1.Login via IDP
In login flow, we use JWKS endpoint to validate id_token during authentication in Mendix.

2.API Authentication using access token
In this flow, we are authenticating the API request by validating access token via introspection/JWKS endpoint

>     * If the IdP does not support token introspection, the OIDC module will assume the access token is a JWT and will validate it using the IdP's public key that is published on the JWKS endpoint.

Reviewer: This is the same mechanism as for vaidating the ID-token:
- the JWT has a signature
- to validate the signature, you need the IdP's public key
- the public key is obtained from the IDP's jwks endpoint.

@MarkvanMents  can i trust you to find the right explanation/wording?

Reviewer: So after Sharishma's comment:

ID-token is always validated based on signature/jwks
Access Token is validated via introspection, if that's not available via signature/jwks.

To parse access tokens, you need to check **Enable Access Token Parsing** when performing [OIDC Client Configuration](#client-configuration).

> From version **VERSION** of the OIDC SSO module, the following claims will be used to assign the user role, depending on the IdP:

Author: I'm not sure about this. We say that the customer still has to write a custom microflow to populate `OIDC.Role` and has to pick it up from from the token.

So are these the values that the custom microflow should be looking for - and shouldn't the information therefore be in the appropriate section of 7.3.2 where we say 

> Your token will contain one of the following:

> * the UUIDs of the user roles in your app which map to the `System.UserRole/ModelGUID` attribute
> * the name of the user role in the app, which can be used to find the `System.UserRole` within the app itself using the `Name` attribute

Reviewer: Yes I think we should add the specifics to each of the paragraphs:
- Section about PIB AT token parsing, should indicate that the scope claim should contain UUIDs of the userroles
- Section about SAM AT Parsing, should indicate that the value of sws.samauth.role.name should contain .... (@Sharishma)
- Section about AzureAD AT token parsing, should contain userrole names as groups? (is this correct @Sharishma ?)

Reviewer: @MarkvanMents , I realise this should go into the 2nd bigger PR probably.

Author: OK - I'll remove this and put it into the bigger PR.

> * SAM – sws.samauth.role.name

Author: What is SAM?

Reviewer: SAM is the name of a Siemens IdP.  See introduction at beginning of section 7.3. 

I think improving that introduction is better than this 'isolated' statement about claims, see https://paper.dropbox.com/doc/OIDC-SSO-API-wTm5msWB1rSjVAWGU9ogQ#:uid=429492466955334364440230&h2=7.3-Access-Token-Parsing

have a look!

Author: Will look.

To parse access tokens, you need to check **Enable Access Token Parsing** when performing [OIDC Client Configuration](#client-configuration).

{{% alert color="info" %}}

> In version **VERSION** and above of the OIDC SSO module you will also find microflows for parsing PIB and Azure AD tokens. These will be released publicly in the future and documented at that time.

Author: Still need version number

5. Configure the user roles in your app to match the roles returned by SAM. End-users will be given the matching role when they sign into the app. If the role in the SAM token is not found in the Mendix app the end-user will be given the role `User`.
6. Save the configuration.

#### 7.2.2 Parsing Access Tokens Using a Custom Microflow

> If you choose to implement your own microflow to parse an access token, the microflow name must start with `CustomATP`, for example `CustomATP_MyTokenParser`. The custom microflow has an `Administration.Account` object as the parameter.

Reviewer 2: @MarkvanMents 
Here there is a change
the microflow name should **contain** 'CustomATP'

Author: ```suggestion
If you choose to implement your own microflow to parse an access token, the microflow name must contain `CustomATP`, for example `CustomATP_MyTokenParser`. The custom microflow has an `Administration.Account` object as the parameter.
```

5. Configure the user roles in your app to match the roles returned by SAM. End-users will be given the matching role when they sign into the app. If the role in the SAM token is not found in the Mendix app the end-user will be given the role `User`.
6. Save the configuration.

#### 7.2.2 Parsing Access Tokens Using a Custom Microflow

> If you choose to implement your own microflow to parse an access token, the microflow name must contain `CustomATP`, for example `CustomATP_MyTokenParser`. The custom microflow has an `Administration.Account` object as the parameter.

Reviewer 2: @MarkvanMents "The custom microflow has an Administration.Account as parameter"
This is also specific to older version

Author: Updated in https://github.com/mendix/docs/pull/5447/commits/9a59311e615ea22250a0fce0073ae9154d5e3d21

> 1. Retrieve the access token of the account.

Reviewer 2: This point is specific to older versions of OIDC SSO

1. Use the access token to determine the roles the user has within your app when signed in using the OIDC module.

1. Convert these roles to a list of objects containing the user role.
    Your token will contain one of the following:

    * the UUIDs of the user roles in your app which map to the `System.UserRole/ModelGUID` attribute
    * the name of the user role in the app, which can be used to find the `System.UserRole` within the app itself using the `Name` attribute

>     From version **VERSION** of the OIDC SSO module, your custom microflow should return a list of objects of type `OIDC.Role`.

Reviewer 2: It will be clear if we add a point saying:
Custom microflow takes access token as input and return a list of objects of type OIDC.Role

1. Retrieve the access token of the account.
1. Use the access token to determine the roles the user has within your app when signed in using the OIDC module.

1. Convert these roles to a list of objects containing the user role.
    Your token will contain one of the following:

    * the UUIDs of the user roles in your app which map to the `System.UserRole/ModelGUID` attribute
    * the name of the user role in the app, which can be used to find the `System.UserRole` within the app itself using the `Name` attribute

>     From version **VERSION** of the OIDC SSO module, your custom microflow should return a list of objects of type `OIDC.Role`.

Author: What is the version number?

Reviewer 2: version number - 2.0.0
